File: 01_introduction_to_algorithms/Golang/BinarySearch.go
package main

import "fmt"

func checkBin(list []int, i int) int {
	low := 0
	high := len(list) - 1
	for low <= high {
		mid := (low + high) / 2
		if list[mid] == i {
			return mid
		}
		if list[mid] < i {
			low = mid + 1
		} else {
			high = mid - 1
		}
	}
	return -1
}

func RecursiveCheckBin(list []int, item int, high, low int) int {
	if high >= low {
		mid := (high + low) / 2

		if list[mid] == item {
			return mid
		} else if list[mid] > item {
			return RecursiveCheckBin(list, item, mid-1, low)
		} else {
			return RecursiveCheckBin(list, item, high, mid+1)
		}

	}
	return -1
}

func main() {
	list := []int{1, 2, 3, 4, 5}
	fmt.Println(checkBin(list, 2))                                          // 0
	fmt.Println(checkBin(list, -1))                                         // -1
	fmt.Println(RecursiveCheckBin([]int{1, 2, 3, 4, 5}, 2, len(list)-1, 0)) // 1
	fmt.Println(RecursiveCheckBin([]int{1, 2, 3, 4, 5}, 0, len(list)-1, 0)) //-1
}

File: 01_introduction_to_algorithms/zig/binary-search.zig
const std = @import("std");
const print = std.debug.print;
const expect = std.testing.expect;

pub fn main() void {
    const my_list = &[_]i8{ 1, 3, 5, 7, 9 };

    print("{?}\n", .{binarySearch(i8, my_list, 3)});
    print("{?}\n", .{binarySearch(i8, my_list, -1)});
}

fn binarySearch(comptime T: type, list: []const T, item: T) ?usize {
    var low: i32 = 0;
    const u_high: u32 = @truncate(list.len);
    var high: i32 = @intCast(u_high - 1);

    return while (low <= high) {
        const mid = @divTrunc((low + high), 2);
        const m: usize = @intCast(mid);
        const guess = list[m];
        if (guess == item) break m;
        if (guess > item) {
            high = mid - 1;
        } else low = mid + 1;
    } else null;
}

test "binarySearch" {
    const my_list = &[_]i8{ 1, 3, 5, 7, 9 };

    var i = binarySearch(i8, my_list, 3);
    try expect(i != null);
    try expect(i.? == 1);

    i = binarySearch(i8, my_list, -1);
    try expect(i == null);
}
